File: 721-AccountsMerge/721-AccountsMerge.java
// Last updated: 8/16/2026, 9:49:33 PM
class Solution {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        int n = accounts.size();
        UnionFind uf = new UnionFind(n);
        // email -> index of acc
        Map<String, Integer> emailToAcc = new HashMap<>();
        for(int i = 0; i < n; i ++) {
            List<String> acc = accounts.get(i);
            // skip the first string (name)
            for(int j = 1; j < acc.size(); j ++) {
                String e = acc.get(j);
                if (emailToAcc.containsKey(e)) {
                    // current 'e' already exist in one account before, union the 2 accounts
                    uf.union(i, emailToAcc.get(e));
                } else {
                    emailToAcc.put(e, i);
                }
            }
        }

        // index of acc -> list of emails
        Map<Integer, List<String>> emailGroup = new HashMap<>();
        emailToAcc.forEach((e, i) -> {
            int leader = uf.find(i);
            // if(emailGroup.containsKey(leader)) {
            //     emailGroup.get(leader).add(e);
            // } else {
            //     emailGroup.put(leader, new ArrayList<>(Arrays.asList(e)) );
            // }
            emailGroup.computeIfAbsent(leader, (x -> new ArrayList<>())).add(e);
        });

        // convert to results format
        List<List<String>> res = new ArrayList<>();
        emailGroup.forEach((i, emails) -> {
            String name = accounts.get(i).get(0);
            Collections.sort(emails);
            List<String> temp = new ArrayList<>();
            temp.add(name);
            temp.addAll(emails);
            res.add(temp);
        });

        return res;
    }
}

public class UnionFind {
    private int[] par;
    private int[] rank;

    public UnionFind(int n) {
        par = new int[n];
        rank = new int[n];

        for (int i = 0; i < n; i ++) {
            par[i] = i;
            rank[i] = 1;
        }
    }

    // find which group 'x' belongs to, take amortized const time
    public int find(int x) {
        int root = x;
        while(root != par[root]) root = par[root];
        // Path compression: compress the path back to root
        // this operatiaon gives amoritized const time
        while(x != root) {
            int next = par[x];
            par[x] = root;
            x = next;
        }
        return root;
    }

    // unify the groups containing x and y
    public void union(int x, int y) {
        int root1 = find(x);
        int root2 = find(y);
        // x and y already in same group
        if (root1 == root2) return;

        // merge smaller group into the larger one
        if (rank[root1] < rank[root2]) {
            par[root1] = root2;
            rank[root2] += rank[root1];
        } else {
            par[root2] = root1;
            rank[root1] += rank[root2];
        }
    }
}

File: 721-AccountsMerge/721-AccountsMerge.py
# Last updated: 8/30/2026, 8:04:51 PM
1class Solution:
2    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
3        # union find
4        # for acc: union all emails with the first email
5
6        parent = {}         # email -> parent email
7        email_to_name = {}  # email -> name
8
9        def find(x):
10            while x != parent[x]:
11                parent[x] = parent[parent[x]]   # path halving
12                x = parent[x]
13            return x
14        
15        def union(x, y):
16            rx, ry = find(x), find(y)
17            parent[rx] = ry
18        
19        
20        for name, *emails in accounts:
21            # 1: every email as own group, and record its name
22            for email in emails:
23                email_to_name[email] = name
24                if email not in parent:
25                    parent[email] = email   # new email: self as root
26            
27            # 2: union every email in account with the first email
28            for email in emails[1:]:
29                union(email, emails[0])
30
31        # 3: group by root email
32        groups = defaultdict(list)  # root email -> [e1, e2, e3, ..] (inclduing root eamil itself)
33        for email in parent:
34            root = find(email)
35            groups[root].append(email)
36        
37        # 4: sort emails in graoups[root], find name by email_to_name[root]
38        res = []
39        for root, emails in groups.items():
40            name = email_to_name[root]
41            res.append( [name] + sorted(emails) )
42        
43        return res
44